tabbycat/api/serializers.py

```diff
@@ -466,13 +466,14 @@ def create(self, validated_data):
         speakers_data = validated_data.pop('speakers', [])
         break_categories = validated_data.pop('break_categories')
 
-        emoji, code_name = pick_unused_emoji()
-        if 'emoji' not in validated_data or validated_data.get('emoji') is None:
-            validated_data['emoji'] = emoji
-        if 'code_name' not in validated_data or validated_data.get('code_name') is None:
-            validated_data['code_name'] = code_name
-
-        if validated_data['emoji'] == '':
+        emoji = pick_unused_emoji(validated_data['tournament'].id)
+        if emoji:
+            if 'emoji' not in validated_data or validated_data.get('emoji') is None:
+                validated_data['emoji'] = emoji[0]
+            if 'code_name' not in validated_data or validated_data.get('code_name') is None:
+                validated_data['code_name'] = emoji[1]
+
+        if validated_data.get('emoji') == '':
```

Comment:
I don't see why we have to make all these changes; shouldn't just adding the `tournament` parameter be sufficient?
```suggestion
        emoji, code_name = pick_unused_emoji(self.context.get('tournament'))
        if 'emoji' not in validated_data or validated_data.get('emoji') is None:
            validated_data['emoji'] = emoji
        if 'code_name' not in validated_data or validated_data.get('code_name') is None:
            validated_data['code_name'] = code_name

        if validated_data['emoji'] == '':
```

Reply:
Without the extra check, this code could still crash if `pick_unused_emoji` returned `None`. My logic was that something as "unimportant" as picking an emoji shouldn't even theoretically be able to crash something as critical as adding a team. And if the emoji allocation code changes in the future (like to allow users to edit the list or something), this issue could silently come back.

Do I revert it anyways?

Reply:
The code here should not crash if that's the case, as both
```py
validated_data['emoji'] = emoji
validated_data['code_name'] = code_name
```
are allowable when `emoji` or `code_name` are `None`.

The proper way to fix it would be for `pick_unused_emoji()` to return `None, None` on an `IndexError`, making the return type `Tuple[Optional[str], Optional[str]]` (rather than `Optional[Tuple[str, str]]`) as what was expected both here and in `participants/admin.py`.

That way, we wouldn't need this extra check for the emoji, and you wouldn't need any changes in `participants/admin.py`.

Reply:
Making it return a tuple was my first thought too, but I didn't want to change the function signature - I only later realized the function is only used twice, so it's really easy. I'll do that.

Re. type hinting, I quite like the policy of adding type hints any time a function gets a new caller - if it's only used once, don't bother, but the first time someone adds an additional call site, type hints should be added in the same PR.    
*Of course I usually forget to enforce this in the projects that I manage, but that's easily solvable with an approval checklist...*